server/src/test/java/io/crate/planner/optimizer/rule/MoveFilterBeneathJoinTest.java

```diff
@@ -0,0 +1,191 @@
+
+        assertThat(filter).isEqualTo(
+            "Filter[(a > 1)]\n" +
+            "  └ HashJoin[(b = c)]\n" +
+            "    ├ HashJoin[(a = b)]\n" +
+            "    │  ├ Collect[doc.t1 | [] | true]\n" +
+            "    │  └ Collect[doc.t2 | [] | true]\n" +
+            "    └ Collect[doc.t3 | [] | true]"
+        );
+
+
+        var rule = new MoveFilterBeneathJoin();
+        Match<Filter> match = rule.pattern().accept(filter, Captures.empty());
+
+        assertThat(match.isPresent()).isTrue();
+        assertThat(match.value()).isEqualTo(filter);
+
+        var result = rule.apply(match.value(),
+                                match.captures(),
+                                planStats,
+                                CoordinatorTxnCtx.systemTransactionContext(),
+                                sqlExpressions.nodeCtx,
+                                Function.identity());
+
+        assertThat(result).isEqualTo(
+            "HashJoin[(b = c)]\n" +
+            "  ├ Filter[(a > 1)]\n" +
+            "  │  └ HashJoin[(a = b)]\n" +
+            "  │    ├ Collect[doc.t1 | [] | true]\n" +
+            "  │    └ Collect[doc.t2 | [] | true]\n" +
+            "  └ Collect[doc.t3 | [] | true]"
+        );
+    }
+
+    @Test
+    public void test_push_filter_beyond_nested_join_when_relation_is_on_lhs() {
+        var joinCondition1 = sqlExpressions.asSymbol("doc.t1.a = doc.t2.b");
+        var join1 = new HashJoin(c1, c2, joinCondition1);
+        var joinCondition2 = sqlExpressions.asSymbol("doc.t2.b = doc.t3.c");
+        var join2 = new HashJoin(c3, join1, joinCondition2);
+        var filter = new Filter(join2, sqlExpressions.asSymbol("doc.t1.a > 1"));
+
+        assertThat(filter).isEqualTo(
+            "Filter[(a > 1)]\n" +
+            "  └ HashJoin[(b = c)]\n" +
+            "    ├ Collect[doc.t3 | [] | true]\n" +
+            "    └ HashJoin[(a = b)]\n" +
+            "      ├ Collect[doc.t1 | [] | true]\n" +
+            "      └ Collect[doc.t2 | [] | true]"
+        );
+
+
+        var rule = new MoveFilterBeneathJoin();
+        Match<Filter> match = rule.pattern().accept(filter, Captures.empty());
+
+        assertThat(match.isPresent()).isTrue();
+        assertThat(match.value()).isEqualTo(filter);
+
+        var result = rule.apply(match.value(),
+                                match.captures(),
+                                planStats,
+                                CoordinatorTxnCtx.systemTransactionContext(),
+                                sqlExpressions.nodeCtx,
+                                Function.identity());
+
+        assertThat(result).isEqualTo(
+            "HashJoin[(b = c)]\n" +
+            "  ├ Collect[doc.t3 | [] | true]\n" +
+            "  └ Filter[(a > 1)]\n" +
+            "    └ HashJoin[(a = b)]\n" +
+            "      ├ Collect[doc.t1 | [] | true]\n" +
+            "      └ Collect[doc.t2 | [] | true]"
+        );
```

Comment:
Do I get it right, that applying rule one more time will push filter further down (like it's shown in the PR description)?
If yes, can we add a second iteration to the test?

Reply:
The simple case where a filter is pushed through a unested join is covered in the previous test.

Reply:
I think the test name was confusing as it reads like @BaurzhanSakhariev thought that it will push it completely down. I've added name suggestions to clarify this.